### PR TITLE
GH-36220: [CI] Run the "Docker Push" step only on the main branch

### LINKED ISSUE #36220: [CI] Run the "Docker Push" step only on the main branch
### Describe the enhancement requested

We want to cache only images that are built on the main branch.

We don't need to cache images that are built on other branches such as branches created by Dependabot.

### Component(s)

Continuous Integration

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

We want to cache only images that are built on the main branch.

We don't need to cache images that are built on other branches such as branches created by Dependabot.

### What changes are included in this PR?

Add a branch name check.

### Are these changes tested?

No.

### Are there any user-facing changes?

No.
* Closes: #36220

### CODE REVIEW

**Review by assignUser — APPROVED** [review]
+1

File: .github/workflows/cpp.yml [hunk 1]
```diff
@@ -114,7 +114,11 @@ jobs:
           ulimit -c unlimited
           archery docker run ${{ matrix.image }}
       - name: Docker Push
-        if: success() && github.event_name == 'push' && github.repository == 'apache/arrow'
+        if: >-
+          success() &&
```
Comment (assignUser) [inline .github/workflows/cpp.yml:118]: We might be able to reduce code duplication a bit by refactoring this into an action (similar to #35896 )

Comment (kou) [reply .github/workflows/cpp.yml:118]: It's a good idea!

Comment (raulcd) [reply .github/workflows/cpp.yml:118]: I've created an issue to track it: https://github.com/apache/arrow/issues/36235

**Pull request merged**